internal/hooks/hooks.go

```diff
@@ -0,0 +1,20 @@
+// Package hooks enables the implementation of hooks for dynamic
+// autocompletion. Hooks should be added to this package and
+// registered in AppendCompletionHooks in completion_hooks.go.
+//
+// A hook must have the following signature:
+//  func (h *Hooks) Name(ctx context.Context, value cty.Value) ([]decoder.Candidate, error)
+// It receives the current value of the attribute and must return
+// a list of completion candidates.
```

Comment:
I would argue this is something that a compile-time error would explain and it would likely be even more accurate. As much as I appreciate the commenting efforts, the risk here is that we end up changing the interface upstream but won't remember to update this comment. Compiler will however always bring it up if types/interfaces don't match.

We could just refer to `decoder.CompletionFunc` here and document that one upstream instead.